### ISSUE TITLE
[Common.BaseWindows] Forecast with a delay after past targets

### ISSUE THREAD

**Issue by omcandido** [Nixtla/neuralforecast#1065]
### Description

I am trying to forecast some horizon `h` after some delay `d` between my past data `p` and the start of `h`, since not all my past data may be available at prediction time.

i.e.:
|------------ p ------------|- d -|----- h ------|
························································································

I believe this is equivalent to the "Forecast Start Shifting" in Darts.

### Use case

This scenario is very common in production when exogenous variables are reported with a delay.
E.g., you forecast demand but your database is only updated every 48h with the actual values. Even if you are forecasting for tomorrow, you have no data about the actual demand for the past 2 days.

**Comment by elephaint** [comment]
Hi, you can use as forecast horizon `h + d` and simply ignore the first `d` steps.

**Comment by omcandido** [comment]
Yes, indeed, but I expect this approach to perform suboptimally in cases where `d` is of similar magnitude to (or even larger than) `h`.

In a 15min interval series I have 96 observations per day. In the example I give of predicting one day ahead with a 2-day delay I am predicting 288 values just to throw away the first 144. This is a conservative example, the ratio of useful predictions quickly decreases as the delay increases.

**Comment by elephaint** [comment]
To avoid inefficiencies you can also shift your target by `d` once and then forecast `h` steps ahead.

**Comment by omcandido** [comment]
Shifting the target in the df? Wouldn't that lead to having an amount `d` of unavailable past points right before the start of the forecast?
If I'm not mistaken, the TS windows are sampled by creating two continuous chunks `X` and `Y` -where `X` has length `input_size` and `Y` has length `h`- so my `X` would contain some of the yet unavailable targets. Like so:
```
|------------p------------|--d--|--h--|
      |-------------X-----------|--Y--|
```

While my desired samples would look like:
```
|------------p------------|--d--|--h--|
|-------------X-----------|     |--Y--|
```

Please correct me if I'm wrong or if you meant shifting the target in some other way.

**Comment by omcandido** [comment]
Thanks for elaborating.

I do agree that each shifted row represents the correct combination of exogenous + targets that I want, but what about the target lags?

I am confused by the fact that you seem to refer to each row as an independent sample (which makes me think that I may have misunderstood how NF creates samples).

I would agree that each row is a sample if we were talking about mlforecast, where the target and the lags are all in the same row. But I understand that this is different in neuralforecast, since given a df with just the target `y` (and `ds` and `unique_id`) all inputs and labels (X and Y in my previous comment) will be created according to the `h` and `input_size` parameters. This is what I tried to say with the illustrations of my previous comment and why I think we have to think about how NF samples internally.

So, let's make things more concrete with an example: 
I have 6 data points with no exogenous variables and I want to forecast with parameters `p=3`, `d=2` and `h=1`:
![imagen](https://github.com/user-attachments/assets/71a891fd-8fc4-46de-a05f-c43cc717abe4)
In the original series, that means I want to predict `y(t+d+1)=y(6)=60` based on the observations that `y(1)=10`, `y(2)=20` and `y(3)=30`.
If I do the shift trick, I do get rid of the delay and my new target `y'(t+1)` is correctly set to `y'(4)=60`, but my original first two timestamps (with values 10 and 20) are now gone. And, according to the documentation on BaseWindows models, the prediction will be made based on `y'(1)=30`, `y'(2)=40` and `y'(3)=40` (which is information that won't be available in production).

Hopefully this clarifies my point.
Apologies if I'm there is some fundamental misconception in my question.

**Comment by elephaint** [comment]
Again, please forget about sampling. I think it confuses you and it's irrelevant in this discussion. 

What's important for you, is that each timestamp gets the correct exogenous + target combination. You can do that by (back)shifting the target or (forward)shifting the exogenous. That's it. That's the only thing you need to do. You will get the correct forecasts, for an horizon `h`. The sampling is not relevant for that - we can sample windows in any way we like after constructing our data like this.

To answer some of your questions / remarks:
1. Yes, you lose the first timestamps. That's because you don't have the exogenous for those timestamps (it would be the exogenous for timestamps `0` and -`1`)
2. I don't understand your last comments. By backshifting we have all the information during prediction. At time `t`, we have information up to and including time `t`, and we make predictions for `t+d`.

**Comment by omcandido** [comment]
But let's assume we have no exogenous: how does shifting `y` yield the proper lags + target combination **during training**?

Assuming a horizon of 1, models are trained to predict `P( y[t+1] | y[:t] )`.
If we shift `y` by `d` timestamps, aren't we just predicting `P( y[t+d+1] | y[:t+d] )`? **But we do not know `y[:t+d]`, `only y[:t]`**. We want to train on `P( y[t+d+1] | y[:t] )`.

**Comment by elephaint** [comment]
Options:

1. No exogenous in existing data and you want to train only on the (at least `d`) lagged target variable.

- Add an exogenous with the appropriate lag
- Use an algorithm that can handle (i) exogenous and (ii) `insample_y = False`
- Model is trained only on the exogenous lagged target variable, ignoring the `y` data.

3. With exogenous.

- Backshift `y` by `d`.
- Forward shift exogenous by `d`.

**Comment by elephaint** [comment]
> I think you mean `exclude_insample_y = True`.

Correct

> Using just a handful of lags as exogenous variables seems to defeat the purpose of using neuralforecast; if I have to pick my relevant lags, I might as well just stick to mlforecast and classical models. There's also a conceptual difference between using a single channel of lags (the current trend in SOTA models) VS using several channels of a given lag frequency each (your suggestion).
>
> The same goes for ignoring the autoregressive features; one wants to use deep learning models to detect patterns in y[:t]. If I rely exclusively on correlations with exogenous variables, I may be throwing away valuable information.

There doesn't have to be a difference between using the target lag ('autoregressive features') and a specified lagged exogenous variable. Depending on the algorithm, these variables may be treated equally (typically in MLP or Transformer models, for example). My solution uses a single channel of lags (in Option 1), and possibly (but not necessarily!) multiple in my second solution (Option 2).

I'm not sure I can help you any further. I've offered multiple solutions that can solve your issue. For what it's worth, in Darts they implement this by shifting the targets / covariaties too, so not sure what it is you keep seeking. 

To summarize, in order of easiest -> hardest:
1. Forecast `d+h`, drop `d`;
2. Forward / backshift exogenous / targets;
3. Use a specific lagged exogenous in conjunction with an algorithm that allows `exclude_insample_y` in case only lagged autoregressives are desired;
4. Implement this functionality in NF.

Of course, feel free to take a stab at 4 yourself :) Not saying we'll not implement this any time in the future, but for now there are easy fixes to get you going.

**Comment by elephaint** [comment]
Great, same feelings here! :)
